servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherSwitchMap.java

```diff
@@ -0,0 +1,391 @@
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            // No need to wrap in ConcurrentSubscription because only the latest RSubscriber interacts with the
+            // tSubscription which is atomically protected by the state variable.
+            tSubscription = requireNonNull(subscription);
+            target.onSubscribe(rSubscription);
+            tSubscription.request(1);
+        }
+
+        @Override
+        public void onNext(@Nullable T t) {
+            final Publisher<? extends R> nextPub = requireNonNull(parent.mapper.apply(t),
+                    () -> "Mapper " + parent.mapper + " returned null");
+            currPublisher = new RSubscriber(currPublisher);
+            toSource(nextPub).subscribe(currPublisher);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            if (currPublisher != null) {
+                try {
+                    if (parent.maxDelayedErrors <= 0) {
+                        currPublisher.dispose(EMPTY_SUBSCRIPTION);
+                    }
+                } finally {
+                    final Throwable cause = outerErrorUpdateState(t);
+                    if (cause != null) {
+                        target.onError(cause);
+                    }
+                }
+            } else {
+                target.onError(t);
+            }
+        }
+
+        @Override
+        public void onComplete() {
+            // If current publisher isn't null defer terminal signals to that publisher, otherwise terminate here.
+            TerminalNotification terminalNotification = complete();
+            if (currPublisher == null || (terminalNotification = outerCompleteUpdateState()) != null) {
+                terminalNotification.terminate(target);
+            }
+        }
+
+        @Nullable
+        private Throwable outerErrorUpdateState(Throwable t) {
+            assert currPublisher != null;
+            t = addPendingError(pendingErrorUpdater, pendingErrorCountUpdater, this, parent.maxDelayedErrors, t);
+            for (;;) {
+                final int cState = currPublisher.state;
+                if (stateUpdater.compareAndSet(currPublisher, cState, setOuterState(cState, OUTER_STATE_ERROR))) {
+                    final int innerState = getInnerState(cState);
+                    return parent.maxDelayedErrors <= 0 && innerState != INNER_STATE_ERROR ||
+                            (parent.maxDelayedErrors > 0 &&
+                            (innerState == INNER_STATE_ERROR || innerState == INNER_STATE_COMPLETE)) ? t : null;
+                }
+            }
+        }
+
+        @Nullable
+        private TerminalNotification outerCompleteUpdateState() {
+            assert currPublisher != null;
+            for (;;) {
+                final int cState = currPublisher.state;
+                if (stateUpdater.compareAndSet(currPublisher, cState, setOuterState(cState, OUTER_STATE_COMPLETE))) {
```

Comment:
As far as I can tell, events from the outer `onError` and `onComplete` can race with events from the inner `Publisher` which may lead to RS faults such as an `onNext` after an `onComplete`. This can obviously be solved via `ConcurrentTerminalSubscription` but I think you're trying to avoid that. I think we need to also make sure that the inner state is not in the emitting state and also in in inner-emit code, check if our outer state happened to transition to a terminal state.

Reply:
Here an example I'm thinking of:
initial state: [outer state idle, inner state idle, maxDelayedErrors = 0]
1. Thread1: enter `RSubscriber.onNext`. Set state to `[outer state idle, inner state emitting]`, break loop, pause before `target.onNext(result);`
2. Thread2: call outer `onError(t)`, 
    - call `currPublisher.dispose(EMPTY_SUBSCRIPTION)` sets state to `[outer state idle, inner state disposed]`
    - call `outerErrorUpdateState(t)` sets state to `[outer state error, inner state disposed]`, `parent.maxDelayedErrors <= 0 && innerState != INNER_STATE_ERROR == true -> return t`
    - call `target.onError(t);`
3. Thread1: wake up concurrently with the still running Thread2 which is about to call `target.onError(t)` and call `target.onNext(result)` concurrently.

It looks like we guard against the inner state being in the critical portion of the `onError` pathway but not the `onNext` pathway.

Presuming this all checks out to be true, it might be easier to reshape the outer `onError` method to call specialized version(s) of the inner `dispose` that store the state for the inner loop to handle once it exits the emitting state.

Reply:
good call, this makes sense and I will fix.